vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+    intermediate_node() = delete;
+    intermediate_node(int x, int y)
+        : loc{x, y} {
+        links.resize(4);
+        for (auto i : NODESIDES)
+            links[i] = false;
+        visited = nullptr;
+        on_route = false;
+        has_pins = false;
+        back_bone = false;
+        segment_id = -1;
+    }
+    intermediate_node(location loc_)
+        : loc(loc_) {
+        links.resize(4);
+        for (auto i : NODESIDES)
+            links[i] = false;
+        visited = nullptr;
+        on_route = false;
+        has_pins = false;
+        back_bone = false;
+        segment_id = -1;
+    }
+};
+
+/*
+ * Main class for FPGA Interchange device to rr_graph flow
+ */
+struct RR_Graph_Builder {
+  public:
+    RR_Graph_Builder(Device::Reader& arch_reader,
+                     const DeviceGrid& grid,
+                     DeviceContext& device_ctx,
+                     const std::vector<t_physical_tile_type>& physical_tile_types,
+                     const vtr::vector<RRSegmentId, t_segment_inf>& segment_inf,
+                     const enum e_base_cost_type base_cost_type)
+        : ar_(arch_reader)
+        , device_ctx_(device_ctx)
+        , segment_inf_(segment_inf)
+        , grid_(grid)
+        , base_cost_type_(base_cost_type)
+        , empty_(device_ctx.arch->strings.intern_string(vtr::string_view(""))) {
+        for (auto& type : physical_tile_types) {
+            tile_type_to_pb_type_[std::string(type.name)] = &type;
+        }
+
+        for (size_t i = 0; i < segment_inf.size(); ++i) {
+            wire_name_to_seg_idx_[segment_inf_[RRSegmentId(i)].name] = i;
+        }
+
+        auto primLib = ar_.getPrimLibs();
+        auto portList = primLib.getPortList();
+
+        for (auto cell_mapping : ar_.getCellBelMap()) {
+            size_t cell_name = cell_mapping.getCell();
+
+            int found_valid_prim = false;
+            for (auto primitive : primLib.getCellDecls()) {
+                bool is_prim = str(primitive.getLib()) == std::string("primitives");
+                bool is_cell = cell_name == primitive.getName();
+
+                bool has_inout = false;
+                for (auto port_idx : primitive.getPorts()) {
+                    auto port = portList[port_idx];
+
+                    if (port.getDir() == INOUT) {
+                        has_inout = true;
+                        break;
+                    }
+                }
+
+                if (is_prim && is_cell && !has_inout) {
+                    found_valid_prim = true;
+                    break;
+                }
+            }
+
+            if (!found_valid_prim)
+                continue;
+
+            for (auto common_pins : cell_mapping.getCommonPins()) {
+                for (auto site_type_entry : common_pins.getSiteTypes()) {
+                    for (auto bel : site_type_entry.getBels()) {
+                        bel_cell_mappings_.emplace(bel);
+                    }
+                }
+            }
```

Comment:
I believe that this `bel_cell_mappins_` is pretty similar to what we have for the arch reader. Do you think this can be generalized and put in the `fpga_interchange_utils`?

Reply:
They are extremely similar, but this code does less. It's possible to create new util function for this.

Reply:
I've moved this code to common fpga_interchange_arch_utils